### PR TITLE
map, flatMap, filter on Metadata

### PR DESCRIPTION
**Pull request by paulp**

### CODE REVIEW

**Comment by paulp** [comment]
Only intended as a basis for discussion, not so much as a finished product.

File: core/src/main/scala/api/metadata.scala [hunk 1]
```diff
@@ -39,9 +16,13 @@ object Example {
  *  A typed value can be obtained for any key.
  */
 sealed class Metadata(val attributes: Vector[Attribute]) extends ShowSelf {
+  md =>
+
   private val untypedMap                         = attributes.foldLeft(Map[Key[_], Any]())(_ + _.pair)
   private def untypedAs[A]()(implicit z: Key[A]) = untypedMap(z).asInstanceOf[A]
 
+  def transformAttributes(f: Vector[Attribute] => Vector[Attribute]): Metadata = new Metadata(f(attributes))
```
Comment (EECOLOR) [inline core/src/main/scala/api/metadata.scala:24]: I like this name

**Comment by EECOLOR** [comment]
I have been staring at the rest. And I am convinced we should add methods as we use them. We could spend ages adding and discussing methods that end up not being used.

I think it's great if we have these methods on stand-by though.

**Comment by EECOLOR** [comment]
Oh, and I like the `Only[A]` concept. That can prove very valuable. Not sure about the methods and their signatures until we have a few use cases

**Comment by paulp** [comment]
I might have to stop for now but I pushed my work. In past work I've found a huge source of bugs is confusion about what will happen with symbolic links. Basically every operation where it makes a difference should be forced to either "follow" or "nofollow" to call the method. That's how I arranged it. It also improves the java.nio.file API by a lot, because there's only one "LinkOption" possible. Talk about overabstraction.

**Comment by paulp** [comment]
Restructuring.

**Pull request closed without merging**